### PR TITLE
linked to judge0 api

### PR DESCRIPTION
**Pull request by Milind712000**
# What's this PR do? (related issue if exists)
created /submission [POST] path
installed firebase-admin sdk
installed morgan to log all requests
Fixes: #23 
# Any context you want to provide on the implementation?
download firebase service key from firebase console and place it as private/firebaseServiceKey.json
this file is to be placed in the server and not be disclosed in public
# Screenshots? (necessary if an UI/UX change)
image

### CODE REVIEW

File: views/submitprob.ejs [hunk 1]
```diff
@@ -26,14 +26,52 @@ int main(){
...
+            <option value="4">C (gcc 7.2.0)</option>
+            <option value="3">Basic (fbc 1.05.0)</option>
+            <option value="2">Bash (4.0)</option>
+            <option value="1">Bash (4.4)</option>
         </select>
-        <button type="button" class="btn btn-primary btn-md">Submit</button>
+        <button type="button" onclick="submitSolution();" class="btn btn-primary btn-md">Submit</button>
     </div>
     </div>
 	<footer>
```
Comment (aashish-ak) [inline views/submitprob.ejs]: Indent properly, you can use VSCode to do this, press `Ctrl + Shift + P` then select `Beautify file`, option.

File: public/js/submitprob.js [hunk 2]
```diff
@@ -10,10 +10,39 @@ firebase.initializeApp(config);
 
 const question_id = window.location.href.split ("?")[1];
 var prob_no = $("#prob-no");
+const data = {
+	"code":"_fill",
+	"qID":"_fill",
+	"language":"_fill"
```
[the remark this answers is not in the document]
Comment (aashish-ak) [reply public/js/submitprob.js]: Have you tested this code with some real question and testcase and tried matching it's response?

File: routes/solution.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    //extract the value from resolved promises
+    arr = arr.map(item => item.val());
+
+    return arr;
+}
+
+const submitSolution = async (req,res,next) => {
+    
+    const qID = req.body.qID;
+    const arr = await getParams(qID);
```
Comment (aashish-ak) [inline routes/solution.js]: change `arr` to `data`

File: public/js/submitprob.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Milind712000) [reply public/js/submitprob.js]: I made an example question with id ('test') to tell if a given num is +ve, -ve or zero and tested it with a cpp for correct , wrong and compilation error
the _fill was just placeholder values that were replaced later

File: .gitignore [hunk 4]
```diff
@@ -1,2 +1,3 @@
 node_modules/
-.idea/
\ No newline at end of file
+.idea/
+private/firebaseServiceKey.json
```
[the remark this answers is not in the document]
Comment (Milind712000) [reply .gitignore]: done

File: views/submitprob.ejs (hunk 1 above)
Comment (Milind712000) [reply views/submitprob.ejs]: done

File: routes/solution.js (hunk 3 above)
Comment (Milind712000) [reply routes/solution.js]: done

**Comment by Milind712000** [comment]
image

image


I have tried a different schema
multiple testfiles
each testfile can contain multiple testcase or a single testcase and time is measured individually for each testfile  
If each test case is stored seperately then we will have to combine them into one file before sending to sntc server or send them individually and then check the total time and compare time or if any particular case failed
the problem with latter is as testcases get larger (10^3 and more) we will generate that many requests per user

File: app.js [hunk 5]
```diff
@@ -2,21 +2,26 @@ var express = require ("express");
 var bodyParser = require ("body-parser");
 var path = require ("path");
 var app = express();
+const logger = require('morgan');
+const solution = require('./routes/solution');
+const question = require('./routes/questionV2');
```
Comment (aashish-ak) [inline app.js]: Why is it questionV2 and not just question?

File: routes/questionV2.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    await qRef.child('Total').set(ProbNum);
+    await qRef.child(ProbNum).set(req.body.ques);
+    //submit the testcases
+    await tRef.child(ProbNum).set(req.body.testcases);
+    // console.log('prob number');
+    // console.log(ProbNum);
+    // console.log('question');
+    // console.log(req.body.ques);
+    // console.log('testcase');
+    // console.log(req.body.testcases);
```
Comment (aashish-ak) [inline routes/questionV2.js]: Remove this commented code.

File: routes/solution.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    data["files"].forEach( (testcase,index) => {
+        console.log("checking testcase : ",index);
+        options.body['stdin'] = testcase["stdin"];
+        options.body['expected_output'] = testcase["stdout"];
+        testAll.push(request(options));
+    });
+
+    const SNTCresponse = await Promise.all(testAll) ;
+
+    return SNTCresponse;
```
Comment (aashish-ak) [inline routes/solution.js]: Change `SNTC` to `Judge0`.

File: views/question.ejs [hunk 8]
```diff
@@ -6,29 +6,12 @@
...
-    grid-column: main;
-}
-.center {
-    display: flex;
-    justify-content: center;
-    align-items: center;
-}
 .box pre{
     white-space: pre-wrap;
 }
```
Comment (aashish-ak) [inline views/question.ejs]: Add frequestly used styling codes in a common css file.

**Comment by Vishal1541** [comment]
> image
> 
> image
> 
> I have tried a different schema
> multiple testfiles
> each testfile can contain multiple testcase or a single testcase and time is measured individually for each testfile
> If each test case is stored seperately then we will have to combine them into one file before sending to sntc server or send them individually and then check the total time and compare time or if any particular case failed
> the problem with latter is as testcases get larger (10^3 and more) we will generate that many requests per user

@Milind712000 Each test file has to be stored separately, we don't have to combine them into a single file. And each test file will be run independently. There won't be 10^3 test files at any point of the time. Even in Codechef platform, the problem setter can put a maximum of 20 test files for each problem. If he has to put 10^3 test cases, then the problem setter will himself put it into a single file. That's his problem. :+1:

**Comment by Milind712000** [comment]
> > image
> > image
> > I have tried a different schema
> > multiple testfiles
> > each testfile can contain multiple testcase or a single testcase and time is measured individually for each testfile
> > If each test case is stored seperately then we will have to combine them into one file before sending to sntc server or send them individually and then check the total time and compare time or if any particular case failed
> > the problem with latter is as testcases get larger (10^3 and more) we will generate that many requests per user
> 
> @Milind712000 Each test file has to be stored separately, we don't have to combine them into a single file. And each test file will be run independently. There won't be 10^3 test files at any point of the time. Even in Codechef platform, the problem setter can put a maximum of 20 test files for each problem. If he has to put 10^3 test cases, then the problem setter will himself put it into a single file. That's his problem.

I meant 10^3 test case only

File: views/question.ejs (hunk 8 above)
Comment (Milind712000) [reply views/question.ejs]: that's the only seperate rule (.box pre) for that page all other are common and stored in grid.css

**Comment by Milind712000** [comment]
new routes:
problemset : http://localhost:3000/question/all
display : http://localhost:3000/question/display?qID=
submit question : http://localhost:3000/question/submit 
submit solution : http://localhost:3000/solution/submit?

image
image
image

File: app.js [hunk 9]
```diff
@@ -1,22 +1,29 @@
...
+
+app.use("/question", question);
+app.use('/solution', solution);
+app.use("/", routes);
+
+app.listen(port, () => {
+    console.log("Server started at port " + port);
 });
 
 exports = module.exports = app
```
Comment (aashish-ak) [inline app.js]: Newline here.

File: public/css/sheet.css [hunk 10]
```diff
@@ -198,4 +198,8 @@ h3
 .caption {
   font-weight: bold;
   font-size: 4rem;
+}
+/* margin at top of bulma tables */
+.table {
+  margin-top: 20px;
 }
```
Comment (aashish-ak) [inline public/css/sheet.css]: Newline here

**Pull request merged**